File: Go/matrices/matrix.go
package matrices

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

// Usual matrix structure.
type Matrix struct {
	// Size of usual matrix.
	Size int
	// Usual matrix itself.
	Matr [][]float64
}

// Creating an instance of a usual matrix and returning a pointer.
func NewMatrix(size int) *Matrix {
	m := new(Matrix)
	m.Size = size
	m.Matr = make([][]float64, size)
	for i := 0; i < size; i++ {
		m.Matr[i] = make([]float64, size)
	}
	return m
}

// File input.
func (m *Matrix) In(line string) {
	strs := strings.Split(line, " ")
	counter := 0

	for i := 0; i < m.Size; i++ {
		for j := 0; j < m.Size; j++ {
			m.Matr[i][j], _ = strconv.ParseFloat(strs[counter], 64)
			counter++
		}
	}
}

// Random input.
func (m *Matrix) RandomIn() {
	for i := 0; i < m.Size; i++ {
		for j := 0; j < m.Size; j++ {
			m.Matr[i][j] = float64(rand.Int31()%100) * rand.Float64()
		}
	}
}

// File output.
func (m *Matrix) Out(f *os.File) {
	f.WriteString("Size: ")
	f.WriteString(fmt.Sprintf("%d\n", m.Size))
	f.WriteString("Average: ")
	f.WriteString(fmt.Sprintf("%f\n", m.GetAverage()))
	f.WriteString("Type: Matrix\n")

	for i := 0; i < m.Size; i++ {
		for j := 0; j < m.Size; j++ {
			f.WriteString(fmt.Sprintf("%f ", m.Matr[i][j]))
		}
		f.WriteString("\n")
	}
}

// Getting average of all the elements.
func (m *Matrix) GetAverage() float64 {
	// Total sum of all the elements.
	var sum float64 = 0

	for i := 0; i < m.Size; i++ {
		for j := 0; j < m.Size; j++ {
			sum += m.Matr[i][j]
		}
	}

	return sum / float64(m.Size*m.Size)
}

File: Rust/src/matrices/matrix.rs
use std::fs::File;
use std::io::Write;
use rand::Rng;
use rand::rngs::ThreadRng;
use crate::matrices::base_matrix::BaseMatrix;

pub struct Matrix {
    pub size: usize,
    pub matr: Vec<Vec<f64>>,
}

impl BaseMatrix for Matrix {
    fn input(&mut self, info: &mut [String], index: &mut usize) {
        self.size = info[*index].parse::<usize>().unwrap();
        *index += 1;
        let matr_info: Vec<String> = info[*index].split(' ').map(|s: &str| s.to_string()).collect();
        *index += 1;

        let mut counter: usize = 0;
        for i in 0..self.size {
            self.matr.push(Vec::new());
            for _j in 0..self.size {
                self.matr[i].push(matr_info[counter].parse::<f64>().unwrap());
                counter += 1;
            }
        }
    }

    fn random_input(&mut self) {
        let mut gen: ThreadRng = rand::thread_rng();
        let _rand_num: usize = gen.gen();

        self.size = gen.gen_range(1..101);
        for i in 0..self.size {
            self.matr.push(Vec::new());
            for _j in 0..self.size {
                self.matr[i].push(gen.gen_range(-1000.0..1000.0));
            }
        }
    }

    fn output(&self, file: &mut File) -> Result<(), std::io::Error> {
        file.write_all(format!("Usual matrix with size of {} and average element {}:\n", self.size, self.get_average()).as_bytes())?;

        for i in 0..self.size {
            for j in 0..self.size {
                file.write_all(format!("{} ", self.matr[i][j]).as_bytes())?;
            }
            file.write_all(b"\n")?;
        }

        Ok(())
    }

    fn get_average(&self) -> f64 {
        let mut sum: f64 = 0_f64;
        for i in 0..self.size {
            for j in 0..self.size {
                sum += self.matr[i][j];
            }
        }
        return sum / self.size as f64 / self.size as f64;
    }
}
